lib/DxcSupport/dxcapi.extval.cpp

```diff
@@ -1,42 +1,453 @@
+#ifndef _WIN32
+#include "dxc/WinAdapter.h"
+#endif
+
+#include "dxc/Support/Global.h" // for hresult handling with DXC_FAILED
+#include "dxc/Support/HLSLOptions.h"
 #include "dxc/Support/WinIncludes.h"
-#include <dxc/Support/Global.h> // for hresult handling with DXC_FAILED
-#include <filesystem>           // C++17 and later
+#include <filesystem> // C++17 and later
+#include <sstream>
 // WinIncludes must come before dxcapi.extval.h
+#include "dxc/Support/FileIOHelper.h"
 #include "dxc/Support/dxcapi.extval.h"
+#include "dxc/Support/dxcapi.impl.h"
+#include "dxc/Support/microcom.h"
+#include "dxc/dxcapi.internal.h"
+#include <iostream>
+
+#include "llvm/ADT/SmallVector.h"
+
+namespace {
+
+// The ExtValidationArgHelper class helps manage arguments for external
+// validation, as well as perform the validation step if needed.
+class ExtValidationArgHelper {
+public:
+  ExtValidationArgHelper() = default;
+  ExtValidationArgHelper(const ExtValidationArgHelper &) = delete;
+  HRESULT initialize(IDxcValidator *NewValidator, LPCWSTR **Arguments,
+                     UINT32 *ArgCount, LPCWSTR TargetProfile = nullptr) {
+    IFR(setValidator(NewValidator));
+    IFR(processArgs(Arguments, ArgCount, TargetProfile));
+    return S_OK;
+  }
+
+  HRESULT doValidation(IDxcOperationResult *CompileResult, REFIID Riid,
+                       void **ValResult) {
+
+    if (!CompileResult)
+      return E_INVALIDARG;
+
+    // No validation needed; just set the result and return.
+    if (!NeedsValidation)
+      return CompileResult->QueryInterface(Riid, ValResult);
+
+    // Get the compiled shader.
+    CComPtr<IDxcBlob> CompiledBlob;
+    IFR(CompileResult->GetResult(&CompiledBlob));
+
+    // If no compiled blob; just return the compile result.
+    if (!CompiledBlob)
+      return CompileResult->QueryInterface(Riid, ValResult);
+
+    // Validate the compiled shader.
+    CComPtr<IDxcOperationResult> TempValidationResult;
+    UINT32 DxcValidatorFlags =
+        DxcValidatorFlags_InPlaceEdit |
+        (RootSignatureOnly ? DxcValidatorFlags_RootSignatureOnly : 0);
+    IFR(Validator->Validate(CompiledBlob, DxcValidatorFlags,
+                            &TempValidationResult));
+
+    // Return the validation result if it failed.
+    HRESULT HR;
+    IFR(TempValidationResult->GetStatus(&HR));
+    if (FAILED(HR)) {
```

Comment:
I see uses of DXC_FAILED and FAILED. Hadn't seen DXC_FAILED before but it looks like they do the same thing.
We should just use one.